org.eclipse.m2e.core/src/org/eclipse/m2e/core/internal/embedder/MavenImpl.java

```diff
@@ -1103,4 +1065,82 @@ public MavenExecutionContext createExecutionContext() {
     return new MavenExecutionContext(this, null, null);
   }
 
+  private static final record SettingsCacheKey(File user, File global) {
+
+  }
+
+  private static final class SettingsCache {
+    private long userSettingsLength;
+
+    private long userSettingsTimestamp;
+
+    private long globalSettingsLength;
+
+    private long globalSettingsTimestamp;
+
+    private SettingsCacheKey key;
+
+    private Settings settings;
+
+    private IComponentLookup lookup;
+
+    public SettingsCache(SettingsCacheKey key, IComponentLookup lookup) {
+      this.key = key;
+      this.lookup = lookup;
+    }
+
+    private synchronized Settings getSettings(boolean forceReload) throws CoreException {
+      File global = key.global();
+      long gs;
+      long gt;
+      if(global != null && global.isFile()) {
+        gs = global.length();
+        gt = global.lastModified();
+      } else {
+        gs = -1;
+        gt = -1;
+      }
+      File user = key.user();
+      long us;
+      long ut;
+      if(user != null && user.isFile()) {
+        us = user.length();
+        ut = user.lastModified();
+      } else {
+        us = -1;
+        ut = -1;
+      }
+      boolean reload = forceReload || settings == null || gs != globalSettingsLength || gt != globalSettingsTimestamp
+          || us != userSettingsLength || ut != userSettingsTimestamp;
+      if(reload) {
+        SettingsBuildingRequest request = new DefaultSettingsBuildingRequest();
+        Properties systemProperties = new Properties();
+        M2EUtils.copyProperties(systemProperties, System.getProperties());
+        request.setSystemProperties(systemProperties);
+        if(global != null && !global.equals(SettingsXmlConfigurationProcessor.DEFAULT_GLOBAL_SETTINGS_FILE)) {
+          request.setGlobalSettingsFile(global);
+        }
+        if(user != null) {
+          request.setUserSettingsFile(user);
+        }
+        try {
+          settings = lookup.lookup(SettingsBuilder.class).build(request).getEffectiveSettings();
+        } catch(SettingsBuildingException ex) {
+          String msg = "Could not read settings.xml, assuming default values";
+          log.error(msg, ex);
+          /*
+           * NOTE: This method provides input for various other core functions, just bailing out would make m2e highly
+           * unusuable. Instead, we fail gracefully and just ignore the broken settings, using defaults.
+           */
+          settings = new Settings();
+        }
+        globalSettingsLength = gs;
```

Comment:
Should we update even if we catch an exception? If we saw an exception, we have to reload anyway, don't we?

Reply:
Yes the idea is unless you change the file parsing will always fail the same way (e.g. invalid content)